File: Javascript/telemetry-system/telemetry-diagnostic-controls.js
var TelemetryClient = require('./telemetry-client.js');
var TelemetryOperation = require('./telemetry-operation.js');

TelemetryDiagnosticControls = function() {

	this._diagnosticChannelConnectionString = function() { return '*111#'; };

	this._telemetryClient = new TelemetryClient();
	this._diagnosticInfo = '';
};

TelemetryDiagnosticControls.prototype = {

	readDiagnosticInfo: function() {
		return this._diagnosticInfo;
	},

	writeDiagnosticInfo: function(newValue) {
		this._diagnosticInfo = newValue;
	},

	checkTransmission: function() {

		this._diagnosticInfo = '';

		this._telemetryClient.disconnect();

		var retryLeft = 3;
		while (this._telemetryClient.onlineStatus() === false && retryLeft > 0) {
			this._telemetryClient.connectStatus(this._diagnosticChannelConnectionString);
			retryLeft -= 1;
		}

		if (this._telemetryClient.onlineStatus() === false) {
			throw 'Unable to connect';
		}

		TelemetryOperation.send(TelemetryOperation.diagnosticMessage());

		this._diagnosticInfo = TelemetryOperation.receive();
	}
};

module.exports = TelemetryDiagnosticControls;

File: Swift/RacingCars/Sources/TelemetrySystem/TelemetryDiagnosticControls.swift
import Foundation

public class TelemetryDiagnosticControls {
    enum Error: Swift.Error {
        case unableToConnect
    }

    private let diagnosticChannelConnectionString = "*111#"
    private let _telemetryClient: TelemetryClient
    private var _diagnosticInfo = ""

    public var diagnosticInfo: String {
        get { _diagnosticInfo }
        set { _diagnosticInfo = newValue }
    }

    public init() {
        _telemetryClient = TelemetryClient()
    }

    public func checkTransmission() throws {
        _diagnosticInfo = ""
        _telemetryClient.disconnect()

        var retryLeft = 3
        while _telemetryClient.onlineStatus == false && retryLeft > 0 {
            try _telemetryClient.connect(to: diagnosticChannelConnectionString)
            retryLeft -= 1
        }

        if _telemetryClient.onlineStatus == false {
            throw Error.unableToConnect
        }

        try _telemetryClient.send(TelemetryClient.diagnosticMessage)
        _diagnosticInfo = _telemetryClient.receive()
    }
}
